### ISSUE TITLE
Impossible to customize the style of a dialog's ::backdrop residing inside a Shadow DOM.

### ISSUE THREAD

**Issue by freshp86** [whatwg/fullscreen#124]
**Note:** Copied from https://github.com/whatwg/html/issues/3601. Please continue discussion here.

More context at https://bugs.chromium.org/p/chromium/issues/detail?id=827397.

### Minimal repro 1
This just showcases that any CSS variable is ignored from ::backdrop elements, even without Shadow DOM usage.
https://jsfiddle.net/1zevfdce/3/

### Minimal repro 2
Showcases the problem when a` <dialog>` resides inside a Shadow DOM.
https://jsfiddle.net/o1trLoqL/2/

Note that the [spec](https://fullscreen.spec.whatwg.org/#::backdrop-pseudo-element) mentions the following

"It does not inherit from any element and is not inherited from. No restrictions are made on what properties apply to this pseudo-element either."

Is that the reason for the current behavior? If yes, should the spec be revised, such that customizing the style of a ::backdrop element is possible even if it resides inside a Shadow DOM? Or is there already a viable workaround?

cc @annevk @foolip @domenic @TakayoshiKochi

**Comment by guest271314** [comment]
Have you tried using "blue" as the property?

**Comment by guest271314** [comment]
By `::backdrop` do you mean the screen behind the `<dialog>` element?

**Comment by guest271314** [comment]
Have you tried without CSS `var()` function?

**Comment by upsuper** [comment]
> Is that the reason for the current behavior?

Yes. CSS variables are propagated via inheritance into descendants, so if a pseudo-element doesn't inherit from anything, CSS variables which are not defined for the pseudo-element directly would have no effect on the pseudo-element.

**Comment by guest271314** [comment]
Both 


```
```
dialog::backdrop {
  background-color: blue;
}
```

and 

````
::backdrop {
  --bg-color: blue;
}

dialog::backdrop {
  background-color: var(--bg-color);
}
```

appear to render expected result at Chromium.

**Comment by guest271314** [comment]
```
dialog::backdrop {
  background-color: blue;
}
```

and 

```

::backdrop {
  --bg-color: blue;
}

dialog::backdrop {
  background-color: var(--bg-color);
}

```
appear to render expected result at Chromium.

**Comment by freshp86** [comment]
@guest271314: True, that seems to work, but it still does not solve the issue of the 2nd minimal repro, which involves Shadow DOM.

**Comment by guest271314** [comment]
Have not tried Polymer. Can the code be composed without using a library?

**Comment by freshp86** [comment]
[Here is a simple example](https://jsfiddle.net/062m2fhv/3/) that puts a `<dialog>` inside a Shadow root, without using Polymer. Note that I am not 100% sure that this is entirely equivalent, since it is adding a shadow root on a `<div>` element, instead of using a custom element.

**Comment by guest271314** [comment]
You can include `<style>` element within Shadow DOM to achieve the same result 

```
let d = document.querySelector('#parent');
let shadowRoot = d.attachShadow({ mode: 'open' });
let style = document.createElement("style");
style.textContent = `::backdrop {
  --bg-color: blue;
}

dialog::backdrop {
  background-color: var(--bg-color);
}`;
let child = document.createElement('dialog');
child.textContent = 'foo';

shadowRoot.appendChild(style);
shadowRoot.appendChild(child);

let b = document.querySelector('button');
b.addEventListener('click', () => {
  child.showModal();
});
```

https://jsfiddle.net/062m2fhv/5/

**Comment by freshp86** [comment]
@guest271314: Thanks for your effort to help find a workaround.

Having said that, this bug is claiming that there is no proper  (for example without using JS) way to style a `<dialog>`'s backdrop that lives within a custom element. Specifically see [these docs](https://developers.google.com/web/fundamentals/web-components/shadowdom#stylefromoutside) (which are not using Polymer), on how the internals of a custom element should be able to be styled **from the outside**.

Your workaround shows that one can programatically create a `dialog` inside a Shadow DOM and, programmatically attach a style to it. It could be potentially useful, but still it seems that the current choice of the spec on how `::backdrop` works, makes it impossible (I would gladly be proven wrong), to style such an element with HTML/CSS code only.

**Comment by guest271314** [comment]
"without using JS" does not appear at OP? Not sure what you mean by "impossible", or how the linked document is related to the HTML standard? What exactly is the the requirement?

**Comment by guest271314** [comment]
> such that one can stamp two instances of the custom element, and have instance A use a different backdrop than instance B

You can use the appropriate `var()` at CSS and select the specific `<dialog>` element https://jsfiddle.net/062m2fhv/7/.

**Comment by guest271314** [comment]
HTML

```
<button>Open blue dialog inside Shadow DOM</button>
<button>Open green dialog inside Shadow DOM</button>
<div id="parent"></div>
```

JavaScript

```
let d = document.querySelector('#parent');
let shadowRoot = d.attachShadow({
  mode: 'open'
});
let style = document.createElement("style");
style.textContent = `::backdrop {
  --bg-color: blue;
  --second-bg-color: green;
}

dialog:nth-of-type(1)::backdrop {
  background-color: var(--bg-color);
}
dialog:nth-of-type(2)::backdrop {
  background-color: var(--second-bg-color);
}
`;
let child = document.createElement('dialog');
child.textContent = 'foo';
let child1 = document.createElement('dialog');
child1.textContent = 'bar';

shadowRoot.appendChild(style);
shadowRoot.appendChild(child);
shadowRoot.appendChild(child1);

var dialogs = shadowRoot.querySelectorAll("dialog");

let b = document.querySelectorAll('button');
b.forEach((button, index) => {
  button.addEventListener('click', () => {
    dialogs[index].showModal();
  });
});
```

**Comment by guest271314** [comment]
Can you clarify the parameters of

> with HTML/CSS code only

?

**Comment by annevk** [comment]
Really need @tabatkins and @fantasai's expertise to figure out what the right model for `::backdrop` is I think.

And the other thing we need to know is that if we make it inherit, what would break...

**Comment by upsuper** [comment]
I wonder why was it made not inherit anything? If there isn't really any compelling reason, I guess we should probably just have it inherit from the element.

**Comment by annevk** [comment]
There wasn't a compelling reason. As I said in the other issue it was to address https://lists.w3.org/Archives/Public/www-style/2012Jun/0441.html and back then we didn't have CSS variables. You don't think there's any observable effects to doing this? I guess most inheriting properties won't have an effect on `::backdrop`.

**Comment by guest271314** [comment]
@annevk How could `::backdrop` inherit from an element within a `document` where the effect is rendered at the entire `document`?

**Comment by upsuper** [comment]
How an element/pseudo-element is rendered has no effect on where it inherits from. It's only a question of where is the element/pseudo-element logically located in the DOM tree.

e.g., `::before` is assumed to be placed at the very beginning inside its originating element, so it inherits from the originating element. But you can always make it cover the whole viewport like `::backdrop` via giving it some proper styles like `position: fixed; top: 0; right: 0; bottom: 0; left: 0;`.

In the `::backdrop` case, I think the difficulty is that there is no logical location in DOM tree to place this pseudo-element. Having said that, I realized that there might be some inconsistency if we have it inherit from the originating element, because `::backdrop` is not supposed to be a child of the element at all. But it's probably not a big problem, since top layer would make the pseudo-element escape from any containing block so it doesn't matter whether it is a children of anything in terms of rendering...

**Comment by guest271314** [comment]
@upsuper From the expected effect perspective the only logical element that can fathom which `::backdrop` could inherit from would be `<html>` or `:root` pseudo class, though the current "inheritance" from `::backdrop` appears to provide a means to utilize specificity as to selectors to display different styles for different `<dialog>` elements, unless missing altogether the gist of the current issue.

**Comment by freshp86** [comment]
What about `dialog::backdrop`? Isn't it clear in this case that the element resides within a `dialog`?

**Comment by guest271314** [comment]
What is the issue with the code at the link? `:host::backdrop` styles are applied. Still not gathering what the issue is.

**Comment by freshp86** [comment]
@guest271314: I've tried to explain the issue multiple times. See minimal repro 2 at the start of this bug.
CSS variables applied to a shadow root host don't seem to apply to a `dialog::backdrop` that resides under that shadow root, which is different than the behavior for any other element under the same shadow root.

**Comment by guest271314** [comment]
Is it not possible to reproduce the result at Firefox?

**Comment by freshp86** [comment]
AFAIK `<dialog>` is not implemented in Firefox, so no.

**Comment by tabatkins** [comment]
We should just say that ::backdrop inherits custom properties from its originating element, even if we explicitly block all other inheritance.

(This *should* be doable in pure CSS by saying that `::backdrop { all: initial; --: inherit; }` applies at the UA stylesheet level; this blocks all inheritance of normal properties, but lets custom properties thru. Unfortunately I haven't defined the `--` property yet. :(

**Comment by upsuper** [comment]
I don't think it makes much sense to distinguish between variables and other properties. That may add implementation complexity for no good reason.

**Comment by guest271314** [comment]
Polymer is not currently being loaded at linked jsfiddle. Can the complete issue be reproduced without using a library?

**Comment by guest271314** [comment]
A workaround which gets the style properties of  host element from `document.styleSheets` which begin with `--`, and replaces any matching existing properties and values within `<style>` element of Shadow DOM of host element with the value defined at the style sheet

```
let d = document.querySelector('#parent');
let shadowRoot = d.attachShadow({
  mode: 'open'
});
let style = document.createElement("style");
let child = document.createElement('dialog');
let props = [];
for (let sheet of document.styleSheets) {
  for (let cssRules of sheet.cssRules) {
    if (document.querySelector(cssRules.selectorText) === d) {
      for (let rule of cssRules.style) {
        if (/--/.test(rule)) {
          props.push([rule, cssRules.style.getPropertyValue(rule)]);
        }
      }
    }
  }
}
style.textContent = `dialog::backdrop{background-color:var(--bg-color)}`;
shadowRoot.prepend(style);

if (props.length) {
  for (let [key, prop] of props)
    style.textContent = style.textContent.replace(`var(${key})`, prop)
}

child.textContent = 'foo';
shadowRoot.appendChild(child);

let b = document.querySelector('button');
b.addEventListener('click', () => {
  child.showModal();
});
```

**Comment by fantasai** [comment]
I agree with @upsuper’s comments. Pseudo-elements generally inherit from there originating element, and since they don't contain any text or other content, it seems unlikely that there's a Web-compat reason for `::backdrop` not to do so as well. (If there is, we can take @tabatkins approach of setting `::backdrop { all: initial; }` in the UA stylesheet.)

**Comment by dbatiste** [comment]
I recently encountered this limitation.   We would typically define our color palette with CSS variables at the document root, and then use those variables throughout components, either assigning to component specific variables or assigning directly to CSS properties.  This is my demo.

https://stackblitz.com/edit/dialog-backdrop-test?file=style.css

**Comment by yisibl** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=996467

**Comment by nuxodin** [comment]
Because of the ever-widening support for dialog elements, they are starting to become popular.
The sooner we fix this problem, the less webcompat issuse will be.

And I could use it.
`::backdrop` is the only place where I have to manually assign the color value.

**Comment by mgenware** [comment]
I'm working on a `<dialog>` element-based web component. One requirement is to allow the user to customize dialog background (backdrop area). I keep running into different obstacles with different approaches.

- `::backdrop` + CSS variable: not working with shadow DOM (tried putting custom css variables under `:root` or `:host` or `::backdrop`, none of them work.
- Export `<dialog>` via CSS shadow parts: `part(dialog)` works but `part(dialog)::backdrop` doesn't have any effects
- Roll my own background layer with a `<div>`: when multiple dialogs are open, the backing `<div>` of the topmost dialog couldn't cover the dialogs underneath it as `<dialog>` elements are in [top layer](https://fullscreen.spec.whatwg.org/#new-stacking-layer).

I ended up with a traditional approach without `<dialog>`.

**Comment by mfreed7** [comment]
> @mfreed7 do you have any thoughts about how we could resolve this? What approach would a normal web component use to allow this customization, and can we copy that approach in the UA stylesheet?

So this relates to this issue: https://github.com/w3c/csswg-drafts/issues/6641. I think it would be good for you to chime in on that issue, bringing back up the `::backdrop` issue.

It sounds (?) like the current guidance from that issue is to use `@property`, which does work for this use case:

```html
<div>
  <template shadowroot=open>
    <dialog>I'm in the top layer, inside shadow DOM</div>
    <style>
      ::backdrop {
        background-color:var(--foo,red);
      }
    </style>
  </template>
</div>
<style>
  @property --foo {
    syntax: '<color>';
    inherits: false;
    initial-value: blue;
  }
</style>
<script>
  document.querySelector('div').shadowRoot.firstElementChild.showModal();
</script>
```

In Chrome, that shows up with a blue `::backdrop`.

...but they acknowledge that this is a clumsy solution with several shortcomings. I thought I'd post it in case it's a way forward for you.

**Comment by fantasai** [comment]
@tabatkins and I don't understand why `::backdrop` can't just inherit from its originating element. If there's no reason for it, can we just update the spec to say that?

See also https://github.com/whatwg/fullscreen/issues/124#issuecomment-379701382

**Comment by mfreed7** [comment]
> @tabatkins and I don't understand why `::backdrop` can't just inherit from its originating element. If there's no reason for it, can we just update the spec to say that?
> 
> See also [#124 (comment)](https://github.com/whatwg/fullscreen/issues/124#issuecomment-379701382)

That sounds reasonable to me, and would also fix the problem with custom properties inside `::backdrop` (@jh3y). This might not be web compatible but that remains to be seen.

**Comment by smaug----** [comment]
@emilio

**Comment by emilio** [comment]
Yeah, I don't recall why we decided on `::backdrop`, maybe @upsuper does? I guess when it was used only for fullscreen it was more of a "global" thing, but now I'm not sure that makes sense.

**Comment by upsuper** [comment]
@emilio I'm not sure we discussed about it before actually. It not inheriting anything was added in https://github.com/whatwg/fullscreen/commit/aa8c3cedd19390f5bfb933bd23ac201b4be02b81 so maybe @annevk can shed some light.

But as I mentioned in https://github.com/whatwg/fullscreen/issues/124#issuecomment-378436978 (and several comments before), I think it makes sense for `::backdrop` to inherit from the originating element.

**Comment by annevk** [comment]
Thanks for finding the commit. That allowed me to find https://lists.w3.org/Archives/Public/public-webapps/2012AprJun/1319.html. I suspect I just went with what seemed simplest to address @fantasai's comments at the time and @tabatkins agreed with the fix in a subsequent email. If everyone is on board with changing it that seems fine.

**Comment by jpzwarte** [comment]
Is this still the place for tracking `::backdrop` inheriting (at least CSS custom properties) from its parent element? If so, what is needed to move this along? We are working on a web-component based design system and are running into this.

**Comment by annevk** [comment]
Looking at https://drafts.csswg.org/css-position-4/#backdrop I think this got fixed as part of moving `::backdrop`'s definition into CSS (i.e., 666438755d2f252be1d1f73d8214ac494489d32c).

If this is not yet universally implemented writing web-platform-tests (or improving existing tests) would help move the implementation process along the most.

Keeping this open until we have good test coverage seems reasonable. Any new issues should be filed against CSS Positioned Layout.

**Comment by keithamus** [comment]
I’ve encountered this issue and found other pseudo elements have the same behaviour, for example `::selection`. It makes things slightly difficult because combining them into one selector with other globals (`:root, ::selection, ::backdrop`) will reject the entire style rule on browsers that don’t support one of those elements, which means the entire style rule needs to be duplicated for each selector.

**Comment by jpzwarte** [comment]
> I’ve encountered this issue and found other pseudo elements have the same behaviour, for example `::selection`. It makes things slightly difficult because combining them into one selector with other globals (`:root, ::selection, ::backdrop`) will reject the entire style rule on browsers that don’t support one of those elements, which means the entire style rule needs to be duplicated for each selector.

Untested, but isn't `:where` a solution here?

```css
:where(:root, ::selection, ::backdrop) {
...
}
```

**Comment by keithamus** [comment]
`:where` certainly helps as it's forgiving; if it aligns well with your browser support matrix. It's well supported enough that that is likely the case, but there are still in-the-wild browsers that don't support it.

**Comment by emilio** [comment]
Pseudo-elements in `:where()` are also not valid. So, not a solution. In any case this is fixed at the spec level. People should file browser bugs (and if possible, write tests).

**Comment by emilio** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1855668 for Firefox.

**Comment by emilio** [comment]
I also wrote a patch + fixed tests, so I guess that's not needed anymore.

**Comment by emilio** [comment]
One thing that came up while fixing tests were tests that were doing:

```
dialog { visibility: hidden }
```

And expecting the backdrop to render. I don't expect that to come up in practice, but if such thing became a backwards-compatibility issue, we'd need to do something like adding this to the HTML style sheet:

```
dialog::backdrop { visibility: visible }
```

Again, I don't think we should do this unless required by compat.

**Comment by wes-goulet** [comment]
Seems like is tracked in chromium with https://bugs.chromium.org/p/chromium/issues/detail?id=827397 (I tested in chrome with the `BackdropInheritOriginating` flag enabled and it worked, CSS var passed to `::backdrop`).

Anyone know if there is a webkit bug for this?  I couldn't find one.

**Comment by keithamus** [comment]
https://bugs.webkit.org/show_bug.cgi?id=263834 is now resolved. 

Is it worth closing this issue out now?

**Comment by annevk** [comment]
@keithamus from that PR it seems like the only coverage for this is [`html/semantics/interactive-elements/the-dialog-element/backdrop-inherits.html`](https://github.com/web-platform-tests/wpt/blob/master/html/semantics/interactive-elements/the-dialog-element/backdrop-inherits.html)? Which doesn't address fullscreen. I suppose it might be okay, but it doesn't seem ideal.

**Comment by keithamus** [comment]
Right of course good point.

**Comment by yisibl** [comment]
Glad to see Chrome will be shipping soon as well: https://groups.google.com/a/chromium.org/g/blink-dev/c/yXTxBfLthzc
